addons/website_forum/tests/test_forum.py

```diff
@@ -198,10 +198,36 @@ def test_vote_crash(self):
 
     def test_vote(self):
         self.post.create_uid.karma = KARMA['ask']
-        self.user_portal.karma = KARMA['upv']
-        self.post.with_user(self.user_portal).vote(upvote=True)
+        self.user_portal.karma = KARMA['dwv']
+        initial_vote_count = self.post.vote_count
+        res = self.post.with_user(self.user_portal).vote(upvote=True)
+
+        self.assertEqual('1', res['user_vote'])
+        self.assertEqual(res['vote_count'], initial_vote_count + 1)
+        self.assertEqual(1, self.post.with_user(self.user_portal).user_vote)
         self.assertEqual(self.post.create_uid.karma, KARMA['ask'] + KARMA['gen_que_upv'], 'website_forum: wrong karma generation of upvoted question author')
 
+        # On voting again with the same value, nothing changes
+        res = self.post.with_user(self.user_portal).vote(upvote=True)
+        self.assertEqual(res['vote_count'], initial_vote_count + 1)
+        self.assertEqual('1', res['user_vote'])
+        self.post.invalidate_cache()
+        self.assertEqual(1, self.post.with_user(self.user_portal).user_vote)
+
+        # On reverting vote, vote cancels
+        res = self.post.with_user(self.user_portal).vote(upvote=False)
+        self.assertEqual(res['vote_count'], initial_vote_count)
+        self.assertEqual('0', res['user_vote'])
+        self.post.invalidate_cache()
+        self.assertEqual(0, self.post.with_user(self.user_portal).user_vote)
+
+        # Everything works from "0" too
+        res = self.post.with_user(self.user_portal).vote(upvote=False)
+        self.assertEqual(res['vote_count'], initial_vote_count - 1)
+        self.assertEqual('-1', res['user_vote'])
+        self.post.invalidate_cache()
+        self.assertEqual(-1, self.post.with_user(self.user_portal).user_vote)
```

Comment:
We could also check the created "vote" records, just to check what was created during the test (notably that we have only 1 vote / post / user, ...). Not required to check them at each step, just a final check maybe ?